### PR TITLE
pipeline/psth.py: new TrialCondition rework

### PR DESCRIPTION
**Pull request by ixcat**

### CODE REVIEW

File: pipeline/psth.py [hunk 1]
```diff
@@ -19,129 +22,229 @@
...
+        return cls.get_func({'trial_condition_desc': trial_condition_desc})()
+
+    @classmethod
+    def get_func(cls, key):
+        self = cls()
+
+        func, args = (self & key).fetch1(
+            'trial_condition_func', 'trial_condition_arg')
+
+        return partial(dict(getmembers(cls))[func], **args)
```
Comment (eywalker) [inline pipeline/psth.py]: you can write this as `getattr`

File: pipeline/psth.py [hunk 2]
```diff
@@ -19,129 +22,229 @@
...
+
+    @classmethod
+    def _get_trials_no_stim(cls, task=None, task_protocol=None, outcome=None,
+                            early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_no_stim', locals())
+
+        return ((experiment.BehaviorTrial
+                 & {'task': task}
```
Comment (eywalker) [inline pipeline/psth.py]: write this as one dictionary instead

File: pipeline/psth.py [hunk 3]
```diff
@@ -19,129 +22,229 @@
...
+                 & {'early_lick': early_lick}
+                 & {'outcome': outcome}) - experiment.PhotostimEvent)
+
+    @classmethod
+    def _get_trials_stim(cls, task=None, task_protocol=None, outcome=None,
+                         early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_stim', locals())
+
+        return ((experiment.BehaviorTrial
```
Comment (eywalker) [inline pipeline/psth.py]: Does restricting by `None` behave well? If not, perform some exclusion.

File: pipeline/psth.py [hunk 4]
```diff
@@ -19,129 +22,229 @@
...
+    -> TrialCondition
+    -> ephys.Unit
+    ---
+    unit_psth=NULL:                             longblob
+    """
     psth_params = {'xmin': -3, 'xmax': 3, 'binsize': 0.04}
 
-    @staticmethod
-    def compute_unit_trial_psth(unit_key, trial_keys):
+    key_source = (ephys.Unit * TrialCondition
```
Comment (eywalker) [inline pipeline/psth.py]: Why are we restricting to the specific conditions? Just create relevant trial conditions and let this populate fully.

File: pipeline/psth.py [hunk 5]
```diff
@@ -19,129 +22,229 @@
...
+                                  'good_noearlylick_right_hit',
+                                  'good_noearlylick_left_miss',
+                                  'good_noearlylick_right_miss']])
 
-        q = (ephys.TrialSpikes() & unit_key & trial_keys)
+    def make(self, key):
+        log.info('UnitPsth.make(): key: {}'.format(key))
+
+        # build unit key,
+        unit = {k: v for k, v in key.items() if k in ephys.Unit.primary_key}
```
Comment (eywalker) [inline pipeline/psth.py]: There should be no need for this to be restricted down to the unit key - just use `key` as is in the restriction.

File: pipeline/psth.py [hunk 6]
```diff
@@ -19,129 +22,229 @@
...
-        spikes = np.concatenate(session_unit_spikes)
+        if len(spikes) == 0:
+            log.warning('no spikes found for key {} - null psth'.format(key))
+            self.insert1(key)
+            return
 
-        xmin, xmax, bins = UnitPsth.psth_params.values()
+        spikes = np.concatenate(spikes)
+
+        xmin, xmax, bins = self.psth_params.values()
```
Comment (eywalker) [inline pipeline/psth.py]: Why is not using the unit psth function?

File: pipeline/psth.py (hunk 3 above)
Comment (ixcat) [reply pipeline/psth.py]: no - need to rework (see 'FIXME' on line 59; will adjust before completion)

File: pipeline/psth.py (hunk 4 above)
Comment (ixcat) [reply pipeline/psth.py]: There is one condition for all units w/o L/R which isn't needed for the plots; concievably (but not yet actually) possibly others; was initially added due to the 'handling None' issue discussed in other comment, but these actually are different issues (TrialCondition logic should handle None's, but this still doesn't mean all conditions will need UnitPSTH)..

can discuss further

File: pipeline/psth.py (hunk 6 above)
Comment (ixcat) [reply pipeline/psth.py]: was thinking this function is not needed with the table present (see NotImplementedErrors in not-yet-ported-or-deleted functions below). Will incorporate or not through the remaining plotting. I suppose no harm in having a function though..

Comment (ixcat) [reply pipeline/psth.py]: implemented via function; whynot.

File: pipeline/psth.py (hunk 5 above)
Comment (ixcat) [reply pipeline/psth.py]: fixed

File: pipeline/psth.py (hunk 4 above)
Comment (ixcat) [reply pipeline/psth.py]: actually some other conditions are needed for plots (e.g. `plot_unit_bilateral_photostim_effect`, will adjust `Null` handling & add appropriate supporting criteria)

File: pipeline/psth.py (hunk 1 above)
Comment (ixcat) [reply pipeline/psth.py]: This didn't seem to 'do' for datajoint classes (seems like dj metaclass problem) e.g. given:

```
class Foo:
    def foo():
        pass
```
vs
`pipeline.UnitPsth`:


```
>>> psth.UnitPsth
<class 'pipeline.psth.UnitPsth'>
>>> Foo
<class 'pipeline.shell.Foo'>
>>> getattr(Foo, 'foo')
<function Foo.foo at 0x12ab3ec80>
>>> getattr(psth.UnitPsth, 'get_func')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/chris/src/djc/datajoint-python/datajoint/user_tables.py", line 45, in __getattribute__
    else super().__getattribute__(name))
AttributeError: type object 'UnitPsth' has no attribute 'get_func'
>>> getattr(psth.UnitPsth(), 'get_func')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'UnitPsth' object has no attribute 'get_func'
```

File: pipeline/psth.py (hunk 2 above)
Comment (ixcat) [reply pipeline/psth.py]: adjusted; now just applying **kwargs

File: pipeline/psth.py (hunk 3 above)
Comment (ixcat) [reply pipeline/psth.py]: handled now (simply applying **kwargs rather than explicit processing)

File: pipeline/psth.py [hunk 7]
```diff
@@ -115,26 +271,30 @@ def compute_psth(session_unit_spikes):
 
         return np.array(psth)
```
Comment (ttngu207) [inline pipeline/psth.py]: yeah we can remove this `compute_unit_trial_psth()` here.

File: pipeline/plot/unit_characteristic_plot.py [hunk 8]
```diff
@@ -129,20 +129,33 @@ def plot_unit_selectivity(probe_insert_key, axs=None):
...
-    bi_stim_trials = experiment.BehaviorTrial * experiment.PhotostimTrial & trial_restrictor & both_alm_stim_res
+    no_stim_cond = (psth.TrialCondition
+                    & {'trial_condition_desc':
+                       'all_noearlylick_both_alm_nostim'}).fetch1('KEY')
+
+    bi_stim_cond = (psth.TrialCondition
+                    & {'trial_condition_desc':
+                       'all_noearlylick_both_alm_stim'}).fetch1('KEY')
 
     units = ephys.Unit & 'unit_quality = "good"'
```
Comment (ttngu207) [inline pipeline/plot/unit_characteristic_plot.py]: `units = ephys.Unit & probe_insert_key &  'unit_quality = "good"'`

File: pipeline/plot/unit_characteristic_plot.py [hunk 9]
```diff
@@ -129,20 +129,33 @@ def plot_unit_selectivity(probe_insert_key, axs=None):
 
 
 def plot_unit_bilateral_photostim_effect(probe_insert_key, axs=None):
```
Comment (eywalker) [inline pipeline/plot/unit_characteristic_plot.py]: the convention is to name the axes argument ax, and not axs. Minor point but would improve coherence with other libraries. In particular, axs should be reserved for the case where an array of Axes object is expected.

File: pipeline/plot/unit_characteristic_plot.py [hunk 10]
```diff
@@ -155,148 +168,202 @@ def plot_unit_bilateral_photostim_effect(probe_insert_key, axs=None):
...
-    axs[0].set_title('Contra-selective Units')
+    sel_c = (ephys.Unit * psth.UnitSelectivity
+             & 'unit_selectivity = "contra-selective"' & probe_insert_key)
 
-    # ipsi-selective units
-    # _plot_stacked_psth_diff(
-    #     (psth.UnitPsth * ipsi_sel_units & ipsi_trials).fetch(order_by='unit_posy desc'),
-    #     (psth.UnitPsth * ipsi_sel_units & contra_trials).fetch(order_by='unit_posy desc'),
-    #     ax=axs[1], vlines=period_starts)
+    psth_is_it = (((psth.UnitPsth & conds_i)
```
Comment (eywalker) [inline pipeline/plot/unit_characteristic_plot.py]: Later on, please add comments to clarify what these are: `PSTH over IPSI trials for IPSI selective units`

File: pipeline/plot/unit_characteristic_plot.py [hunk 11]
```diff
@@ -155,148 +168,202 @@ def plot_unit_bilateral_photostim_effect(probe_insert_key, axs=None):
...
+                            vlines=period_starts, flip=True)
 
-    ipsi_sel_units = ephys.Unit * psth.UnitSelectivity & 'unit_selectivity = "ipsi-selective"'
-    contra_sel_units = ephys.Unit * psth.UnitSelectivity & 'unit_selectivity = "contra-selective"'
+    axs[1].set_title('Ipsi-selective Units')
+    axs[1].set_ylabel('Unit (by depth)')
+    axs[1].set_xlabel('Time to go (s)')
+
+
+def plot_ave_contra_ipsi_psth(probe_insert_key, axs=None):
```
Comment (eywalker) [inline pipeline/plot/unit_characteristic_plot.py]: I know I'm being picky but I fine `ave` as an shorthand for `average` to be a bit awkward - would prefer `avg`

File: pipeline/plot/unit_characteristic_plot.py (hunk 8 above)
Comment (eywalker) [inline pipeline/plot/unit_characteristic_plot.py]: Missing restriction by `probe_insert_key`

File: pipeline/psth.py (hunk 1 above)
Comment (eywalker) [reply pipeline/psth.py]: I see - that's weird and to be assessed but in the future.

File: pipeline/plot/unit_characteristic_plot.py (hunk 11 above)
Comment (ttngu207) [reply pipeline/plot/unit_characteristic_plot.py]: ah, agreed

File: pipeline/plot/unit_characteristic_plot.py (hunk 9 above)
Comment (ttngu207) [reply pipeline/plot/unit_characteristic_plot.py]: even in the case of ax containing multiple axes

Comment (eywalker) [reply pipeline/plot/unit_characteristic_plot.py]: Oops, did I misplace my comment? I meant this only for a plot that was producing single panel plot, not those that have multiple axes.

File: pipeline/plot/unit_characteristic_plot.py (hunk 10 above)
Comment (ixcat) [reply pipeline/plot/unit_characteristic_plot.py]: fixed

File: pipeline/plot/unit_characteristic_plot.py (hunk 8 above)
Comment (ixcat) [reply pipeline/plot/unit_characteristic_plot.py]: this is the condition key; not actual 'expansion' into trials

relatedly TrialCondition yields a query and not a fetch so if it were needed, one would:

`(psth.TrialCondition.get_trials(...) & probe_insert_key).fetch()`

perhaps TrialCondition functions should be renamed... hmm.

Comment (ixcat) [reply pipeline/plot/unit_characteristic_plot.py]: side note the other, possibly confusing usage when casually viewed here, usage of trial condition where restriction is applied occurs in UnitPsth, where it is actually an upstream filter on the condition:

```        # expand TrialCondition to trials,
        trials = TrialCondition.get_trials(
            (TrialCondition & key).fetch1('trial_condition_desc'))
```

File: pipeline/psth.py (hunk 7 above)
Comment (ixcat) [reply pipeline/psth.py]: fixed in `c018e278`

File: pipeline/plot/unit_characteristic_plot.py (hunk 8 above)
Comment (ixcat) [reply pipeline/plot/unit_characteristic_plot.py]: fixed in `844c234`

**Pull request merged**